src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs

```diff
@@ -0,0 +1,146 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using Nito.AsyncEx;
+using Raven.Client.Documents.Replication.Messages;
+using Raven.Client.Exceptions.Sharding;
+using Raven.Server.Config;
+using Raven.Server.Documents.Replication;
+using Raven.Server.Documents.Replication.Incoming;
+using Raven.Server.Documents.Replication.ReplicationItems;
+using Raven.Server.Documents.TcpHandlers;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Json;
+using Sparrow.Server;
+using Voron;
+
+namespace Raven.Server.Documents.Sharding.Handlers
+{
+    public class ShardedIncomingReplicationHandler : AbstractIncomingReplicationHandler<TransactionContextPool, TransactionOperationContext>
+    {
+        private readonly ShardedDatabaseContext.ShardedReplicationContext _parent;
+        private readonly ReplicationQueue _replicationQueue;
+
+        public ShardedIncomingReplicationHandler(TcpConnectionOptions tcpConnectionOptions, ShardedDatabaseContext.ShardedReplicationContext parent,
+            JsonOperationContext.MemoryBuffer buffer, ReplicationLatestEtagRequest replicatedLastEtag, ReplicationQueue replicationQueue)
+            : base(tcpConnectionOptions, buffer, parent.Server, parent.DatabaseName, replicatedLastEtag, parent.Context.DatabaseShutdown, parent.Server.ContextPool)
+        {
+            _parent = parent;
+            _replicationQueue = replicationQueue;
+            _attachmentStreamsTempFile = new StreamsTempFile("ShardedReplication" + Guid.NewGuid(), false);
+        }
+
+        protected override void EnsureNotDeleted(string nodeTag) => _parent.EnsureNotDeleted(_parent.Server.NodeTag);
+
+        protected override void OnFailed(Exception exception)
+        {
+        }
+
+        protected override void HandleHeartbeatMessage(TransactionOperationContext jsonOperationContext, BlittableJsonReaderObject blittableJsonReaderObject)
+        {
```

Comment:
add DevelopmentHelper here with appropriate description, we need to think what to do here. In regular handler this is a place where we are tackling the optimization issue when e.g. when we have master-master replication